lib/actors/actors.nit

```diff
@@ -0,0 +1,104 @@
+# This file is part of NIT (http://www.nitlanguage.org).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# Abstraction of the actors concepts
+module actors is
+	new_annotation actor
+end
+
+import pthreads::concurrent_collections
+intrude import pthreads
+
+# Abstraction of an actor
+# It has a mailbox, can receive and process messages asynchronously
+abstract class Actor
+
+	# Type of the proxied class or the working class
+	# (depending on theaded or scheduled actors)
+	type V: Object
+
+	# The instance used to do the real work
+	# i.e. the real working class
+	var instance: V
+end
+
+# The promise of a value which will be set asynchronously
+class Future[E]
+	# Value promised by `self`
+	var value: nullable E = null
+
+	# Mutex for synchronisation
+	protected var mutex = new Mutex
+
+	# Condition variable for synchronisation
+	protected var cond: nullable PthreadCond = null
+
+	# Can be used to check if the value is available without waiting
+	protected var is_done = false
+
+	# Set the value and signal so that, someone waiting for `value` can retrieve it
+	fun set_value(value: E) do
+		mutex.lock
+		is_done = true
+		self.value = value
+		var cond = self.cond
+		if cond != null then cond.signal
+		mutex.unlock
+	end
+
+	# Return immediatly if `value` is set, or block waiting for `value` to be set
+	fun join: E do
+		mutex.lock
+		if not is_done then
+			var cond = new PthreadCond
+			self.cond = cond
+			cond.wait(mutex)
+		end
+		mutex.unlock
+		return value
+	end
+end
+
+# A mailbox in which you can receive message
+# Basically a `ConcurrentList` renamed for actors purposes
+class Mailbox[E]
+	super ConcurrentList[E]
+end
+
+# A Blocking queue implemented from a `ConcurrentList`
+class BlockingQueue[E]
+	super ConcurrentList[E]
+
+	# Used to block or signal on waiting threads
+	private var cond = new PthreadCond
+
+	# Adding the signal to release eventual waiting thread(s)
```

Comment:
Enfait j'ai mis `thread(s)` parce que le manuel de `pthread_cond_signal()` dit la chose suivante :
**The pthread_cond_signal() function shall unblock at least  one  of  the
       threads  that  are blocked on the specified condition variable cond (if
       any threads are blocked on cond).**
Et donc aparement il y a un risque que plusieurs threads soient réveillés avec `signal`. Pour le moment la `BlockingQueue` est utilisée dans des cas où réveiller plusieurs threads est inutile :
* Dans `threaded_actors.nit`, les mailboxes des acteurs sont des `BlockingQueue` Pour que le thread attende de recevoir un message avant de travailler, donc en pratique il n'y a qu'un thread qui attend dans ce cas là
* Dans `scheduled_actors.nit`, c'est le `Scheduler` qui utilise des `BloquingQueue` pour gérer les acteurs qui ont du travail à faire et  les threads disponibles. Dans ce cas aussi, il n'y a que le thread du `Scheduler` qui attend dessus.